src/client/auth/aws.rs

```diff
@@ -0,0 +1,414 @@
+              content-type;\
+              host;\
+              x-amz-date;\
+              {token_signed_header}\
+              x-mongodb-gs2-cb-flag;\
+              x-mongodb-server-nonce\
+            ",
+            token_signed_header = token_signed_header,
+        );
+
+        let body = "Action=GetCallerIdentity&Version=2011-06-15";
+        let hashed_body = hex::encode(Sha256::digest(body.as_bytes()));
+
+        let nonce = base64::encode(server_nonce);
+
+        #[rustfmt::skip]
+        let request = format!(
+            "\
+             POST\n\
+             /\n\n\
+             content-length:43\n\
+             content-type:application/x-www-form-urlencoded\n\
+             host:{host}\n\
+             x-amz-date:{date}\n\
+             {token}\
+             x-mongodb-gs2-cb-flag:n\n\
+             x-mongodb-server-nonce:{nonce}\n\n\
+             {signed_headers}\n\
+             {hashed_body}\
+             ",
+            host = host,
+            date = date_str,
+            token = token,
+            nonce = nonce,
+            signed_headers = signed_headers,
+            hashed_body = hashed_body,
+        );
+
+        let hashed_request = hex::encode(Sha256::digest(request.as_bytes()));
+
+        let small_date = date.format("%Y%m%d").to_string();
+
+        let region = if host == "sts.amazonaws.com" {
+            "us-east-1"
+        } else {
+            let parts: Vec<_> = host.split('.').collect();
+            parts.get(1).copied().unwrap_or("us-east-1")
+        };
+
+        #[rustfmt::skip]
+        let string_to_sign = format!(
+            "\
+             AWS4-HMAC-SHA256\n\
+             {full_date}\n\
+             {small_date}/{region}/sts/aws4_request\n\
+             {hashed_request}\
+            ",
+            full_date = date_str,
+            small_date = small_date,
+            region = region,
+            hashed_request = hashed_request,
+        );
+
+        let first_hmac_key = format!("AWS4{}", self.secret_key);
+        let k_date = hmac(first_hmac_key, &small_date)?;
+        let k_region = hmac(k_date, region)?;
+        let k_service = hmac(k_region, "sts")?;
+        let k_signing = hmac(k_service, "aws4_request")?;
+
+        let signature_bytes = hmac(k_signing, string_to_sign)?;
+        let signature = hex::encode(signature_bytes);
+
+        #[rustfmt::skip]
+        let auth_header = format!(
+            "\
+             AWS4-HMAC-SHA256 \
+             Credential={access_key}/{small_date}/{region}/sts/aws4_request, \
+             SignedHeaders={signed_headers}, \
+             Signature={signature}\
+            ",
+            access_key = self.access_key,
+            small_date = small_date,
+            region = region,
+            signed_headers = signed_headers,
+            signature = signature
+        );
+
+        Ok(auth_header)
+    }
+}
+
+fn hmac(key: impl AsRef<[u8]>, input: impl AsRef<[u8]>) -> Result<impl AsRef<[u8]>> {
```

Comment:
theres already a `ScramVersion::hmac` method in `scram.rs`. Would it be possible to share code between the two so that we don't have separate hmac functions for different auth mechanisms?

Reply:
does hmac 0.8.x not work for us?

Reply:
The 0.8.x API seems to be designed in way that's not very conducive to making functions generic over the type of Hmac. I think it's possible to get working, but it's a pretty big hassle. They changed the API to split all of the stuff in `Mac` to be half a dozen or so different traits, and they all depend on each other in different ways. Upgrading `hmac` to 0.8.1 without changing the code yields 115 compiler errors. Adding `NewMac` (the trait that now contains `new` and `new_varkey`) bounds to all the parameters currently requiring `Mac` increases that to 116 instead of getting rid of any errors. I tried adding all the traits that are required and got up to 171 compiler errors. The changes are complex enough that I think it's probably a bit out of scope for this ticket. We can file another one and decide how much effort we want to put into it; it honestly might end up being less work to just duplicate a bit of code instead of making it generic. 

More generally, we probably want to figure out some sort of inform policy for when we want to update dependencies; a decent number of them have minor or bugfix releases out newer than the version we specify, so it might be worth deciding "we update all dependencies to the latest bugfix release on the minor version we're using whenever we do a bugfix release of the driver, and we update minor versions whenever we do a minor version release of the driver" or something like that. Right now, we're just updating only when there's a security bug or we need a new feature, and we probably don't want to be constantly updating all of our dependencies as soon as they publish new releases, but there's probably a middle ground between those that would balance the costs of churn with the risks of using outdated versions.

Reply:
Yeah defining such a policy sounds like a good idea. Once we have that in place we can revisit updating `hmac`.